perl-lib/OESS/lib/OESS/FWDCTL/Master.pm

```diff
@@ -2015,17 +2034,50 @@ sub addVlan {
     # proceed. Circuits with a state of scheduled shall not be added.
     if ($details->{'state'} eq 'scheduled') {
         $self->{'logger'}->info("Elapsed time addVlan: " . tv_interval( $start, [gettimeofday]));
-        &$success_callback({status => $result, event_id => $event_id});
+        &$success_callback({status => $result});
     }
 
     $self->{'circuit_status'}->{$circuit_id} = OESS_CIRCUIT_UP;
 
+    my $minisem = scalar keys %dpids;
+    my $err     = undef;
+
     foreach my $dpid (keys %dpids){
-        $self->send_message_to_child($dpid,{action => 'add_vlan', circuit_id => $circuit_id}, $event_id);
-    }
+        $self->{'fwdctl_events'}->{'topic'} = "OF.FWDCTL.Switch." . sprintf("%x", $dpid);
+        $self->{'fwdctl_events'}->add_vlan(
+            circuit_id     => $circuit_id,
+            async_callback => sub {
+                my $res = shift;
+
+                # Once $minisem is zero all responses have been
+                # received. Use this opportunity to check for any
+                # errors that may have occurred.
+                $minisem = $minisem - 1;
+
+                if (defined $res->{'error'}) {
+                    $err = $res->{'error'};
+                    $self->{'logger'}->error($err);
+                }
 
-    $self->{'logger'}->info("Elapsed time addVlan: " . tv_interval( $start, [gettimeofday]));
-    &$success_callback({status => $result, event_id => $event_id});
+                if ($minisem == 0) {
+                    if (defined $err) {
+                        foreach my $id (keys %dpids) {
+                            $self->{'fwdctl_events'}->{'topic'} = "OF.FWDCTL.Switch." . sprintf("%x", $dpid);
+                            $self->{'fwdctl_events'}->remove_vlan(circuit_id     => $circuit_id,
+                                                                  async_callback => sub {
+                                                                      $self->{'logger'}->error("Removed circuit $circuit_id from $dpid.");
+                                                                  });
+                        }
+
+                        $self->{'logger'}->error("Failed to add VLAN. Elapsed time: " . tv_interval($start, [gettimeofday]));
+                        &$error_callback({error => $err});
+                    }
+
+                    $self->{'logger'}->info("Added VLAN. Elapsed time: " . tv_interval($start, [gettimeofday]));
+                    &$success_callback({status => $res->{'results'}->{'status'}});
```

Comment:
If there are multiple switches involved, this doesn't appear to wait for the other devices to send its result?